RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,832 @@
+```
+
+A `let` interrupting the `let! ... and! ...` block:
+
+```fsharp
+ce {
+    let! x = foo
+    let  z = x / 3 ❌
+    and! y = bar
+    return x + y + z
+ }
+```
+
+A `let` after the `and!`s but before `return`:
+
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    let z = y * 2 ❌
+    return x + y + z
+ }
+// Example.fsx(4,5): error FS3243: Expecting 'and!', 'anduse!' or 'return' but saw something else. Applicative computation expressions must be of the form 'let! <pat1> = <expr2> and! <pat2> = <expr2> and! ... and! <patN> = <exprN> return <exprBody>'.
+```
+
+A `yield` instead of a `return`:
+
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    yield x + y ❌
+ }
+```
+
+Multiple `return`s:
+
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    return x + y
+    return (x + y) * 2 ❌
+ }
+```
+
+Other CE keywords anywhere in the expression:
+
+```fsharp
+ce {
+    let! x = foo
+    and! y = bar
+    do! webRequest x y ❌
+    return x + y
+ }
+// Example.fsx(4,5): error FS3243: Expecting 'and!', 'anduse!' or 'return' but saw something else. Applicative computation expressions must be of the form 'let! <pat1> = <expr2> and! <pat2> = <expr2> and! ... and! <patN> = <exprN> return <exprBody>'.
+```
+
+### Rationale for strong syntax constraints
+
+This syntax may sound very constrained, but it is for good reason. The structure imposed by this rule forces the CE to be in a canonical form ([McBride & Paterson](http://www.staff.city.ac.uk/~ross/papers/Applicative.html)):
+
+> Any expression built from the Applicative combinators can be transformed to a canonical form in which a single pure function is "applied" to the effectful parts in depth-first order:  
+`pure f <*> arg1 <*> ... <*> argN`  
+This canonical form captures the essence of Applicative programming: computations have a fixed structure, given by the pure function, and a sequence of subcomputations, given by the effectful arguments.
+
+In our case, the expression to the right of `return` (i.e. `pure`) becomes the body of a lambda, whose parameters are introduced by the `let! ... and! ...` preceding it.
+
+Similarly, the canonical form of `let! ... and! ... return ...` in F# makes should make it clear that what we are really doing it calling the function given to `return` with the arguments introduced by `let! ... and! ...`, but in a special context determined by the CE builder.
+
+Despite requiring the canonical form, there are still many ways to build more complex and useful expressions from this syntax. The rest of this section will detail these.
+
+## Desugaring
+
+Computation expressions are provided meaning via [translation to method calls on a builder class](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-expressions#creating-a-new-type-of-computation-expression).
```

Comment:
Is it worth explicitly mentioning the expected types of the builder methods (as in the table in the [existing documentation for CEs)](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-expressions#creating-a-new-type-of-computation-expression), either here or elsewhere, or do you think that's unnecessary?

(The type of Apply is given in the "Why applicatives?" section, but that seems to be almost in passing.)

Reply:
Good point! That has been sat in my notes for weeks and I'd just kind of presumed it was in the RFC somewhere already.

Thanks!